lib/trust_level.rb

```diff
@@ -3,33 +3,60 @@
 class InvalidTrustLevel < StandardError; end
 
 class TrustLevel
+  def self.[](level)
+    raise InvalidTrustLevel if !valid?(level)
 
-  class << self
+    level
+  end
 
-    def [](level)
-      raise InvalidTrustLevel if !valid?(level)
-      level
-    end
+  def self.levels
+    @levels ||= Enum.new(:newuser, :basic, :member, :regular, :leader, start: 0)
+  end
 
-    def levels
-      @levels ||= Enum.new(:newuser, :basic, :member, :regular, :leader, start: 0)
-    end
+  def self.valid?(level)
+    valid_range === level
+  end
 
-    def valid?(level)
-      valid_range === level
-    end
+  def self.valid_range
+    (0..4)
+  end
 
-    def valid_range
-      (0..4)
-    end
+  def self.compare(current_level, level)
+    (current_level || 0) >= level
+  end
+
+  def self.name(level)
+    I18n.t("js.trust_levels.names.#{levels[level]}")
+  end
 
-    def compare(current_level, level)
-      (current_level || 0) >= level
+  def self.calculate(user, use_previous_trust_level: false)
+    # First, use the manual locked level
+    if user.manual_locked_trust_level.present?
+      return user.manual_locked_trust_level
     end
 
-    def name(level)
-      I18n.t("js.trust_levels.names.#{levels[level]}")
+    # Then consider the group locked level (or the previous trust level)
+    granted_trust_level = user.group_granted_trust_level
+
+    if use_previous_trust_level && previous_trust_level = find_previous_trust_level(user)
+      if !granted_trust_level || previous_trust_level > granted_trust_level
+        granted_trust_level = previous_trust_level
+      end
     end
+
+    granted_trust_level
   end
 
+  private
+
+  def self.find_previous_trust_level(user)
+    user_history = UserHistory
+      .where(action: UserHistory.actions[:change_trust_level])
+      .where(target_user_id: user.id)
+      .order(created_at: :desc)
+      .last
+    return if !user_history
+
+    user_history.new_value.to_i if user_history.new_value.present?
+  end
```

Comment:
I think we could have it be just one sql query without any conditionals like this

It will return `0` when there's no previous trust level which is correct since it's the level you start with anyways.

```suggestion
  def self.find_previous_trust_level(user)
    UserHistory
      .where(action: UserHistory.actions[:change_trust_level])
      .where(target_user_id: user.id)
      .order(created_at: :desc)
      .first
      .pluck(:new_value)
      .to_i
  end
```